# Set initialization file
https://cccma.gitlab.io/classic/makeInputFiles.html

Initialize the model for CLASSIC (CLASS+CTEM) run. Use the Paul's SnowMIP data to set the first soil layers proporties
and then extrapolate with the satellite data. 

In [1]:
# Env: sc2_v0

import xarray as xr
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import proplot as pplt # New plot library (https://proplot.readthedocs.io/en/latest/)
pplt.rc['savefig.dpi'] = 300 # 1200 is too big! #https://proplot.readthedocs.io/en/latest/basics.html#Creating-figures

In [2]:
exp = 'Ref'
site = 'oas'
site_paul = 'Old_Aspen'
site_ex = 'CA-Oas'
# site_ex = 'AU-Tum'

path_in = '/home/lalandmi/eccc/classic-develop/inputFiles/FLUXNETsites/'+site_ex # example file
path_out = '/home/lalandmi/eccc/classic-develop/inputFiles/SnowMIP/'+site

In [3]:
ds = xr.open_dataset(path_in+'/'+site_ex+'_init.nc')
ds.load()

<xarray.Dataset>
Dimensions:            (tile: 1, lat: 1, lon: 1, icp1: 6, layer: 20, ic: 5,
                        icc: 12, iccp1: 13, iccp2: 14, slope: 8)
Coordinates:
  * ic                 (ic) int32 1 2 3 4 5
  * icc                (icc) int32 1 2 3 4 5 6 7 8 9 10 11 12
  * iccp1              (iccp1) int32 1 2 3 4 5 6 7 8 9 10 11 12 13
  * iccp2              (iccp2) int32 1 2 3 4 5 6 7 8 9 10 11 12 13 14
  * icp1               (icp1) int32 1 2 3 4 5 6
  * lat                (lat) float64 53.63
  * layer              (layer) int32 1 2 3 4 5 6 7 8 ... 13 14 15 16 17 18 19 20
  * lon                (lon) float64 -106.2
  * slope              (slope) int32 1 2 3 4 5 6 7 8
  * tile               (tile) int32 1
Data variables: (12/101)
    ALBS               (tile, lat, lon) float64 0.7647
    ALIC               (tile, icp1, lat, lon) float64 0.0 0.29 0.0 0.0 0.0 0.0
    ALVC               (tile, icp1, lat, lon) float64 0.0 0.05 0.0 0.0 0.0 0.0
    CLAY               (tile, layer, lat, lon) float64 15.16 20.08 ... 23.29
    CMAS               (tile, ic, lat, lon) float64 0.0 20.0 0.0 0.0 0.0
    Cmossmas           (tile, lat, lon) float64 0.0
    ...                 ...
    stemmass_s         (tile, icc, lat, lon) float64 0.0 0.0 0.0 ... 0.0 0.0 0.0
    stmhrlos           (tile, icc, lat, lon) float64 0.0 0.0 0.0 ... 0.0 0.0 0.0
    tcoldm             (lat, lon) float32 99.0
    twarmm             (lat, lon) float32 -99.0
    tymaxlai           (tile, icc, lat, lon) float64 0.0 0.0 0.0 ... 0.0 0.0 0.0
    nh4_mass           (tile, iccp1, lat, lon) float32 0.0 0.0 0.0 ... 0.0 0.0
Attributes:
    title:          CLASSIC initialization file created from: /home/acrp001/m...
    creation_date:  20190523
    history:        Wed Jun 12 17:19:01 2019: ncks -4 ./rsfile.nc ./rsfile.nc
    NCO:            4.4.2
    row_bounds:     1 1

In [4]:
ds.soilcmas.sum()

<xarray.DataArray 'soilcmas' ()>
array(0.)

## BERMS Old Aspen, Saskatchewan, Canada

Even-aged stand of aspen with a thick hazelnut understorey, naturally regenerated after
fire in 1919. 10 cm layer of organic litter and peat over a sandy clay loam.

| Short name | oas |
|:-----------|:----|
| Location | 53.63ºN, 106.20ºW |
| Elevation | 600 m |
| Canopy height | 21 m |
| Leaf area index | winter stem area ~1, summer 3.7 – 5.2 |
| Snow-free albedo | 0.14 |
| Simulation period | 1 October 1997 to 30 September 2010 |
| Temperature/humidity measurement height | 37 m |
| Wind measurement height | 38 m |
| Reference | Bartlett et al. (2007) |

In [5]:
lat = 53.62889
lon = -106.19779

# If the site description is not enough we can get the input data from gridded satellite datasets
path = '/home/lalandmi/Dropbox/data/CLASSIC/'
SoilGrids250m_CLAY = xr.open_dataset(path+'/soil/SoilGrids250m_CLAY_0.05deg.nc')
SoilGrids250m_ORGM = xr.open_dataset(path+'/soil/SoilGrids250m_ORGM_0.05deg.nc')
SoilGrids250m_SAND = xr.open_dataset(path+'/soil/SoilGrids250m_SAND_0.05deg.nc')
SoilGrids250m_SDEP = xr.open_dataset(path+'/soil/SoilGrids250m_SDEP_0.05deg.nc')
NCAR_SOCI = xr.open_dataset(path+'/soil/NCAR_SOCI_0.5deg.nc')
c4_fraction_1deg = xr.open_dataarray(path+'/vegetation/c4_fraction_1deg.nc')

path_paul = '/home/lalandmi/Dropbox/data/SnowMIP/Paul' 
ds_init_Paul = xr.open_dataset(path_paul+'/'+site_paul+'_60_ESMSnowMIP.nc').load()

### Set the layer coords to the center soil layer

In [6]:
layer_c = []

for i in range(len(ds.DELZ)):
    if i == 0:
        layer_c.append(ds.DELZ.cumsum().values[i]/2)
    else:
        layer_c.append(ds.DELZ.cumsum().values[i-1] + ds.DELZ.values[i]/2)
        
with xr.set_options(keep_attrs=True):
    ds = ds.assign_coords(layer=ds.layer*0+layer_c)

ds.layer

<xarray.DataArray 'layer' (layer: 20)>
array([5.00e-02, 1.50e-01, 2.50e-01, 3.50e-01, 4.50e-01, 5.50e-01, 6.50e-01,
       7.50e-01, 8.50e-01, 9.50e-01, 1.10e+00, 1.35e+00, 1.70e+00, 2.15e+00,
       2.90e+00, 4.90e+00, 8.90e+00, 1.89e+01, 4.14e+01, 5.89e+01])
Coordinates:
  * layer    (layer) float64 0.05 0.15 0.25 0.35 0.45 ... 4.9 8.9 18.9 41.4 58.9
Attributes:
    standard_name:  ground layers

In [7]:
ds.DELZ.cumsum()

<xarray.DataArray 'DELZ' (layer: 20)>
array([ 0.1,  0.2,  0.3,  0.4,  0.5,  0.6,  0.7,  0.8,  0.9,  1. ,  1.2,
        1.5,  1.9,  2.4,  3.4,  6.4, 11.4, 26.4, 56.4, 61.4])
Coordinates:
  * layer    (layer) float64 0.05 0.15 0.25 0.35 0.45 ... 4.9 8.9 18.9 41.4 58.9

In [8]:
layer_c = []

for i in range(len(ds_init_Paul.DELZ)):
    if i == 0:
        layer_c.append(ds_init_Paul.DELZ.cumsum().values[i]/2)
    else:
        layer_c.append(ds_init_Paul.DELZ.cumsum().values[i-1] + ds_init_Paul.DELZ.values[i]/2)
        
with xr.set_options(keep_attrs=True):
    ds_init_Paul = ds_init_Paul.assign_coords(layer=ds_init_Paul.layer*0+layer_c)

ds_init_Paul.layer

<xarray.DataArray 'layer' (layer: 3)>
array([0.05 , 0.225, 2.225])
Coordinates:
  * layer    (layer) float64 0.05 0.225 2.225
Attributes:
    standard_name:  ground layers

In [9]:
ds_init_Paul.DELZ.cumsum()

<xarray.DataArray 'DELZ' (layer: 3)>
array([0.1 , 0.35, 4.1 ])
Coordinates:
  * layer    (layer) float64 0.05 0.225 2.225

### Combine Paul's data with satellite data

#### SAND

In [10]:
ds_init_Paul.SAND.squeeze().values

array([-2. , 48.8, 53.6])

In [11]:
SoilGrids250m_SAND.sel(lat=lat, lon=lon, method='nearest').SAND.values

array([31.031921, 37.93344 , 41.008335, 42.932327, 42.932327, 42.932327,
       44.295258, 44.295258, 44.295258, 44.295258, 44.019623, 44.019623,
       44.019623, 44.019623, 44.019623, 44.019623, 44.019623, 44.019623,
       44.019623, 44.019623], dtype=float32)

In [12]:
SAND = ds_init_Paul.where(ds_init_Paul.SAND >= 0).interp(layer=ds.layer, method='linear').SAND.squeeze()
SAND = SAND.interpolate_na(dim='layer', method='nearest', fill_value="extrapolate")

# Deals only with the case of the first layer is peat
# (needs to be adapted if more peat layers)
if ds_init_Paul.SAND.squeeze().values[0] == -2:
    SAND.values[0] = -2 

# Average the last layers with satellite data
SAND.values[4:] = (SAND.values[4:] + SoilGrids250m_SAND.sel(lat=lat, lon=lon, method='nearest').SAND.values[4:])/2
SAND

<xarray.DataArray 'SAND' (layer: 20)>
array([-2.        , 48.86      , 48.86      , 49.1       , 46.13616364,
       46.25616364, 47.05762878, 47.17762878, 47.29762878, 47.41762878,
       47.4598114 , 47.7598114 , 48.1798114 , 48.7198114 , 48.7198114 ,
       48.7198114 , 48.7198114 , 48.7198114 , 48.7198114 , 48.7198114 ])
Coordinates:
    lon      float64 253.8
    lat      float64 53.63
    tile     int32 1
  * layer    (layer) float64 0.05 0.15 0.25 0.35 0.45 ... 4.9 8.9 18.9 41.4 58.9
Attributes:
    units:      %
    long_name:  Percentage sand content

#### CLAY

In [13]:
ds_init_Paul.CLAY.squeeze().values

array([ 0. , 24.2, 22.1])

In [14]:
SoilGrids250m_CLAY.sel(lat=lat, lon=lon, method='nearest').CLAY.values

array([15.810381, 20.604921, 23.0375  , 24.032326, 24.032326, 24.032326,
       24.198334, 24.198334, 24.198334, 24.198334, 23.847364, 23.847364,
       23.847364, 23.847364, 23.847364, 23.847364, 23.847364, 23.847364,
       23.847364, 23.847364], dtype=float32)

In [15]:
CLAY = ds_init_Paul.where(ds_init_Paul.SAND >= 0).interp(layer=ds.layer, method='linear').CLAY.squeeze()
CLAY = CLAY.interpolate_na(dim='layer', method='nearest', fill_value="extrapolate")

# Deals only with the case of the first layer is peat
# (needs to be adapted if more peat layers)
if ds_init_Paul.SAND.squeeze().values[0] == -2:
    CLAY.values[0] = 0

# Average the last layers with satellite data
CLAY.values[4:] = (CLAY.values[4:] + SoilGrids250m_CLAY.sel(lat=lat, lon=lon, method='nearest').CLAY.values[4:])/2
CLAY

<xarray.DataArray 'CLAY' (layer: 20)>
array([ 0.        , 24.17375   , 24.17375   , 24.06875   , 23.99803787,
       23.94553787, 23.97604187, 23.92354187, 23.87104187, 23.81854187,
       23.56430721, 23.43305721, 23.24930721, 23.01305721, 23.01305721,
       23.01305721, 23.01305721, 23.01305721, 23.01305721, 23.01305721])
Coordinates:
    lon      float64 253.8
    lat      float64 53.63
    tile     int32 1
  * layer    (layer) float64 0.05 0.15 0.25 0.35 0.45 ... 4.9 8.9 18.9 41.4 58.9
Attributes:
    units:      %
    long_name:  Percentage clay content

#### ORGM

In [16]:
ds_init_Paul.ORGM.squeeze().values

array([77.5,  0.8,  4.1])

In [17]:
SoilGrids250m_ORGM.sel(lat=lat, lon=lon, method='nearest').ORGM.values

array([29.33908  , 12.62931  ,  5.       ,  2.2413793,  2.2413793,
        2.2413793,  1.5660921,  1.5660921,  1.5660921,  1.5660921,
        1.9971266,  1.9971266,  1.9971266,  1.9971266,  1.9971266,
        1.9971266,  1.9971266,  1.9971266,  1.9971266,  1.9971266],
      dtype=float32)

In [18]:
ORGM = ds_init_Paul.where(ds_init_Paul.SAND >= 0).interp(layer=ds.layer, method='linear').ORGM.squeeze()
ORGM = ORGM.interpolate_na(dim='layer', method='nearest', fill_value="extrapolate")

# Deals only with the case of the first layer is peat
# (needs to be adapted if more peat layers)
if ds_init_Paul.SAND.squeeze().values[0] == -2:
    ORGM.values[0] = 0

# Average the last layers with satellite data
ORGM.values[4:] = (ORGM.values[4:] + SoilGrids250m_ORGM.sel(lat=lat, lon=lon, method='nearest').ORGM.values[4:])/2
ORGM

<xarray.DataArray 'ORGM' (layer: 20)>
array([0.        , 0.84125   , 0.84125   , 1.00625   , 1.70631463,
       1.78881463, 1.53367107, 1.61617107, 1.69867107, 1.78117107,
       2.12043829, 2.32668829, 2.61543829, 2.98668829, 2.98668829,
       2.98668829, 2.98668829, 2.98668829, 2.98668829, 2.98668829])
Coordinates:
    lon      float64 253.8
    lat      float64 53.63
    tile     int32 1
  * layer    (layer) float64 0.05 0.15 0.25 0.35 0.45 ... 4.9 8.9 18.9 41.4 58.9
Attributes:
    units:      %
    long_name:  Percentage organic matter content

In [23]:
# These values can be attributed with the global grid if not available
SDEP = SoilGrids250m_SDEP.sel(lat=lat, lon=lon, method='nearest').SDEP.values.item(0) # Soil permeable depth (m)
SOCI = NCAR_SOCI.sel(lat=lat, lon=lon, method='nearest').SOCI.values.item(0) # Soil color inde
c4_fraction = c4_fraction_1deg.sel(lat=lat, lon=lon, method='nearest').values.item(0) # If grass


# Those can be distributed along the soil layers if available
CLAY = CLAY.values
SAND = SAND.values
ORGM = ORGM.values
# If the sum does not reach 100 %, the left over will be attributed to silt

# ! List of CLASS-level PFTs
classpfts = ['NdlTr', 'BdlTr', 'Crops', 'Grass', 'BdlSh']
FCAN = {'NdlTr': 0. , 'BdlTr': 1., 'Crops': 0., 'Grass': 0., 'BdlSh': 0.} # max 1
# Maybe add a little fraction of trees (as it's bordered by coniferous -> TODO)

# ! List of CTEM PFTs
# ! **Note: 'BdlDCoTr' should be specified before 'BdlDDrTr' due to some code in competition.
ctempfts = ['NdlEvgTr', 'NdlDcdTr', 'BdlEvgTr', 'BdlDCoTr', 'BdlDDrTr', 'CropC3', 'CropC4', 'GrassC3', 'GrassC4', 'Sedge', 'BdlEvgSh', 'BdlDCoSh']
fcancmx = {'NdlEvgTr': 0., 'NdlDcdTr': 0., 'BdlEvgTr': 0., 'BdlDCoTr': 1., 'BdlDDrTr': 0., 'CropC3': 0., 'CropC4': 0., 
           'GrassC3': 0., 'GrassC4': 0., 'Sedge': 0., 'BdlEvgSh': 0., 'BdlDCoSh': 0.} # Max 1
# Check differences between C3 and C4 at Cdp? (TODO)

if sum(FCAN.values()) > 1: raise Exception("The sum of FCAN values needs to be lower than 1.")
if sum(fcancmx.values()) > 1: raise Exception("The sum of fcanmax values needs to be lower than 1.")

print('SDEP = ' + str(SDEP))
print('\nSOCI = ' + str(SOCI))
print('\nCLAY = ' + str(CLAY))
print('\nSAND = ' + str(SAND))
print('\nORGM = ' + str(ORGM))
print('\nGrass C4 fraction = ' + str(c4_fraction))

SDEP = 35.869998931884766

SOCI = 19.0

CLAY = [ 0.         24.17375    24.17375    24.06875    23.99803787 23.94553787
 23.97604187 23.92354187 23.87104187 23.81854187 23.56430721 23.43305721
 23.24930721 23.01305721 23.01305721 23.01305721 23.01305721 23.01305721
 23.01305721 23.01305721]

SAND = [-2.         48.86       48.86       49.1        46.13616364 46.25616364
 47.05762878 47.17762878 47.29762878 47.41762878 47.4598114  47.7598114
 48.1798114  48.7198114  48.7198114  48.7198114  48.7198114  48.7198114
 48.7198114  48.7198114 ]

ORGM = [0.         0.84125    0.84125    1.00625    1.70631463 1.78881463
 1.53367107 1.61617107 1.69867107 1.78117107 2.12043829 2.32668829
 2.61543829 2.98668829 2.98668829 2.98668829 2.98668829 2.98668829
 2.98668829 2.98668829]

Grass C4 fraction = 0.00570000009611249


### Set lat/lon

In [24]:
with xr.set_options(keep_attrs=True):
    ds = ds.assign_coords(lat=(ds.lat*0+lat))
    ds = ds.assign_coords(lon=(ds.lon*0+lon))
ds

<xarray.Dataset>
Dimensions:            (tile: 1, lat: 1, lon: 1, icp1: 6, layer: 20, ic: 5,
                        icc: 12, iccp1: 13, iccp2: 14, slope: 8)
Coordinates:
  * ic                 (ic) int32 1 2 3 4 5
  * icc                (icc) int32 1 2 3 4 5 6 7 8 9 10 11 12
  * iccp1              (iccp1) int32 1 2 3 4 5 6 7 8 9 10 11 12 13
  * iccp2              (iccp2) int32 1 2 3 4 5 6 7 8 9 10 11 12 13 14
  * icp1               (icp1) int32 1 2 3 4 5 6
  * lat                (lat) float64 53.63
  * layer              (layer) float64 0.05 0.15 0.25 0.35 ... 18.9 41.4 58.9
  * lon                (lon) float64 -106.2
  * slope              (slope) int32 1 2 3 4 5 6 7 8
  * tile               (tile) int32 1
Data variables: (12/101)
    ALBS               (tile, lat, lon) float64 0.7647
    ALIC               (tile, icp1, lat, lon) float64 0.0 0.29 0.0 0.0 0.0 0.0
    ALVC               (tile, icp1, lat, lon) float64 0.0 0.05 0.0 0.0 0.0 0.0
    CLAY               (tile, layer, lat, lon) float64 15.16 20.08 ... 23.29
    CMAS               (tile, ic, lat, lon) float64 0.0 20.0 0.0 0.0 0.0
    Cmossmas           (tile, lat, lon) float64 0.0
    ...                 ...
    stemmass_s         (tile, icc, lat, lon) float64 0.0 0.0 0.0 ... 0.0 0.0 0.0
    stmhrlos           (tile, icc, lat, lon) float64 0.0 0.0 0.0 ... 0.0 0.0 0.0
    tcoldm             (lat, lon) float32 99.0
    twarmm             (lat, lon) float32 -99.0
    tymaxlai           (tile, icc, lat, lon) float64 0.0 0.0 0.0 ... 0.0 0.0 0.0
    nh4_mass           (tile, iccp1, lat, lon) float32 0.0 0.0 0.0 ... 0.0 0.0
Attributes:
    title:          CLASSIC initialization file created from: /home/acrp001/m...
    creation_date:  20190523
    history:        Wed Jun 12 17:19:01 2019: ncks -4 ./rsfile.nc ./rsfile.nc
    NCO:            4.4.2
    row_bounds:     1 1

### Set PFTs

In [25]:
for i, pft in enumerate(classpfts):
    ds.FCAN[0, i, 0, 0] = FCAN[pft]
    print(pft + ' -> ' + str(FCAN[pft]*100) + ' %')
    
ds.FCAN[0, :, 0, 0]
# 6th PFT: Bareground (need to be specified if CLASS used without CTEM)

NdlTr -> 0.0 %
BdlTr -> 100.0 %
Crops -> 0.0 %
Grass -> 0.0 %
BdlSh -> 0.0 %


<xarray.DataArray 'FCAN' (icp1: 6)>
array([0., 1., 0., 0., 0., 0.])
Coordinates:
  * icp1     (icp1) int32 1 2 3 4 5 6
    lat      float64 53.63
    lon      float64 -106.2
    tile     int32 1
Attributes:
    units:      -
    long_name:  Annual maximum fractional coverage of modelled area (read in ...

In [26]:
for i, pft in enumerate(ctempfts):
    ds.fcancmx[0, i, 0, 0] = fcancmx[pft]
    print(pft + ' -> ' + str(fcancmx[pft]*100) + ' %')
    
ds.fcancmx[0, :, 0, 0]

NdlEvgTr -> 0.0 %
NdlDcdTr -> 0.0 %
BdlEvgTr -> 0.0 %
BdlDCoTr -> 100.0 %
BdlDDrTr -> 0.0 %
CropC3 -> 0.0 %
CropC4 -> 0.0 %
GrassC3 -> 0.0 %
GrassC4 -> 0.0 %
Sedge -> 0.0 %
BdlEvgSh -> 0.0 %
BdlDCoSh -> 0.0 %


<xarray.DataArray 'fcancmx' (icc: 12)>
array([0., 0., 0., 1., 0., 0., 0., 0., 0., 0., 0., 0.])
Coordinates:
  * icc      (icc) int32 1 2 3 4 5 6 7 8 9 10 11 12
    lat      float64 53.63
    lon      float64 -106.2
    tile     int32 1
Attributes:
    units:      -
    long_name:  PFT fractional coverage per grid cell

### Set soil color

In [27]:
with xr.set_options(keep_attrs=True):
    ds['SOCI'][0, 0, 0] = SOCI
ds.SOCI

<xarray.DataArray 'SOCI' (tile: 1, lat: 1, lon: 1)>
array([[[19.]]])
Coordinates:
  * lat      (lat) float64 53.63
  * lon      (lon) float64 -106.2
  * tile     (tile) int32 1
Attributes:
    units:      index
    long_name:  Soil colour index

### Set permeable depth

In [28]:
with xr.set_options(keep_attrs=True):
    ds['SDEP'][0, 0, 0] = SDEP
ds.SDEP

<xarray.DataArray 'SDEP' (tile: 1, lat: 1, lon: 1)>
array([[[35.86999893]]])
Coordinates:
  * lat      (lat) float64 53.63
  * lon      (lon) float64 -106.2
  * tile     (tile) int32 1
Attributes:
    units:      m
    long_name:  Soil permeable depth

### Check maximum level before bedrock
https://gitlab.com/jormelton/classic/-/blob/develop/src/modelStateDrivers.f90?ref_type=heads#L968

In [29]:
ds.DELZ

<xarray.DataArray 'DELZ' (layer: 20)>
array([ 0.1,  0.1,  0.1,  0.1,  0.1,  0.1,  0.1,  0.1,  0.1,  0.1,  0.2,
        0.3,  0.4,  0.5,  1. ,  3. ,  5. , 15. , 30. ,  5. ])
Coordinates:
  * layer    (layer) float64 0.05 0.15 0.25 0.35 0.45 ... 4.9 8.9 18.9 41.4 58.9
Attributes:
    units:      m
    long_name:  Ground layer thickness

In [30]:
i = 0
while ds.DELZ.cumsum()[i] < SDEP:
    i += 1
    if i > ds.DELZ.shape[0]-1:
        print('The permable depth is greater than the model levels')
        break
        
maxlevel = i + 1 # first level of bedrock (next to the one containing SDEP)
maxlevel

19

In [31]:
ds.DELZ.cumsum()

<xarray.DataArray 'DELZ' (layer: 20)>
array([ 0.1,  0.2,  0.3,  0.4,  0.5,  0.6,  0.7,  0.8,  0.9,  1. ,  1.2,
        1.5,  1.9,  2.4,  3.4,  6.4, 11.4, 26.4, 56.4, 61.4])
Coordinates:
  * layer    (layer) float64 0.05 0.15 0.25 0.35 0.45 ... 4.9 8.9 18.9 41.4 58.9

In [32]:
ds.DELZ.cumsum()[maxlevel]

<xarray.DataArray 'DELZ' ()>
array(61.4)
Coordinates:
    layer    float64 58.9

In [33]:
ds.DELZ.cumsum()[:maxlevel]

<xarray.DataArray 'DELZ' (layer: 19)>
array([ 0.1,  0.2,  0.3,  0.4,  0.5,  0.6,  0.7,  0.8,  0.9,  1. ,  1.2,
        1.5,  1.9,  2.4,  3.4,  6.4, 11.4, 26.4, 56.4])
Coordinates:
  * layer    (layer) float64 0.05 0.15 0.25 0.35 0.45 ... 2.9 4.9 8.9 18.9 41.4

In [34]:
ds.DELZ.cumsum()[maxlevel:]

<xarray.DataArray 'DELZ' (layer: 1)>
array([61.4])
Coordinates:
  * layer    (layer) float64 58.9

### Set soil contents and flags
Note: flags are only set in the SAND variable (-3 bedrock / -2 peatland)

In [35]:
with xr.set_options(keep_attrs=True):
    # Set values until bedrock
    ds.SAND[0, :maxlevel, 0, 0] = SAND[:maxlevel]
    ds.CLAY[0, :maxlevel, 0, 0] = CLAY[:maxlevel]
    ds.ORGM[0, :maxlevel, 0, 0] = ORGM[:maxlevel]
    
    # Set bedrock values
    ds.SAND[0, maxlevel:, 0, 0] = -3 # flag for bedrock
    ds.CLAY[0, maxlevel:, 0, 0] = 0.
    ds.ORGM[0, maxlevel:, 0, 0] = 0.

In [36]:
# Current values
ds.SAND[0, :, 0, 0]

<xarray.DataArray 'SAND' (layer: 20)>
array([-2.        , 48.86      , 48.86      , 49.1       , 46.13616364,
       46.25616364, 47.05762878, 47.17762878, 47.29762878, 47.41762878,
       47.4598114 , 47.7598114 , 48.1798114 , 48.7198114 , 48.7198114 ,
       48.7198114 , 48.7198114 , 48.7198114 , 48.7198114 , -3.        ])
Coordinates:
    lat      float64 53.63
  * layer    (layer) float64 0.05 0.15 0.25 0.35 0.45 ... 4.9 8.9 18.9 41.4 58.9
    lon      float64 -106.2
    tile     int32 1
Attributes:
    units:      %
    long_name:  Percentage sand content

In [37]:
ds.CLAY[0, :, 0, 0]

<xarray.DataArray 'CLAY' (layer: 20)>
array([ 0.        , 24.17375   , 24.17375   , 24.06875   , 23.99803787,
       23.94553787, 23.97604187, 23.92354187, 23.87104187, 23.81854187,
       23.56430721, 23.43305721, 23.24930721, 23.01305721, 23.01305721,
       23.01305721, 23.01305721, 23.01305721, 23.01305721,  0.        ])
Coordinates:
    lat      float64 53.63
  * layer    (layer) float64 0.05 0.15 0.25 0.35 0.45 ... 4.9 8.9 18.9 41.4 58.9
    lon      float64 -106.2
    tile     int32 1
Attributes:
    units:      %
    long_name:  Percentage clay content

In [38]:
ds.ORGM[0, :, 0, 0]

<xarray.DataArray 'ORGM' (layer: 20)>
array([0.        , 0.84125   , 0.84125   , 1.00625   , 1.70631463,
       1.78881463, 1.53367107, 1.61617107, 1.69867107, 1.78117107,
       2.12043829, 2.32668829, 2.61543829, 2.98668829, 2.98668829,
       2.98668829, 2.98668829, 2.98668829, 2.98668829, 0.        ])
Coordinates:
    lat      float64 53.63
  * layer    (layer) float64 0.05 0.15 0.25 0.35 0.45 ... 4.9 8.9 18.9 41.4 58.9
    lon      float64 -106.2
    tile     int32 1
Attributes:
    units:      %
    long_name:  Percentage organic matter content

In [39]:
ds

<xarray.Dataset>
Dimensions:            (tile: 1, lat: 1, lon: 1, icp1: 6, layer: 20, ic: 5,
                        icc: 12, iccp1: 13, iccp2: 14, slope: 8)
Coordinates:
  * ic                 (ic) int32 1 2 3 4 5
  * icc                (icc) int32 1 2 3 4 5 6 7 8 9 10 11 12
  * iccp1              (iccp1) int32 1 2 3 4 5 6 7 8 9 10 11 12 13
  * iccp2              (iccp2) int32 1 2 3 4 5 6 7 8 9 10 11 12 13 14
  * icp1               (icp1) int32 1 2 3 4 5 6
  * lat                (lat) float64 53.63
  * layer              (layer) float64 0.05 0.15 0.25 0.35 ... 18.9 41.4 58.9
  * lon                (lon) float64 -106.2
  * slope              (slope) int32 1 2 3 4 5 6 7 8
  * tile               (tile) int32 1
Data variables: (12/101)
    ALBS               (tile, lat, lon) float64 0.7647
    ALIC               (tile, icp1, lat, lon) float64 0.0 0.29 0.0 0.0 0.0 0.0
    ALVC               (tile, icp1, lat, lon) float64 0.0 0.05 0.0 0.0 0.0 0.0
    CLAY               (tile, layer, lat, lon) float64 0.0 24.17 ... 23.01 0.0
    CMAS               (tile, ic, lat, lon) float64 0.0 20.0 0.0 0.0 0.0
    Cmossmas           (tile, lat, lon) float64 0.0
    ...                 ...
    stemmass_s         (tile, icc, lat, lon) float64 0.0 0.0 0.0 ... 0.0 0.0 0.0
    stmhrlos           (tile, icc, lat, lon) float64 0.0 0.0 0.0 ... 0.0 0.0 0.0
    tcoldm             (lat, lon) float32 99.0
    twarmm             (lat, lon) float32 -99.0
    tymaxlai           (tile, icc, lat, lon) float64 0.0 0.0 0.0 ... 0.0 0.0 0.0
    nh4_mass           (tile, iccp1, lat, lon) float32 0.0 0.0 0.0 ... 0.0 0.0
Attributes:
    title:          CLASSIC initialization file created from: /home/acrp001/m...
    creation_date:  20190523
    history:        Wed Jun 12 17:19:01 2019: ncks -4 ./rsfile.nc ./rsfile.nc
    NCO:            4.4.2
    row_bounds:     1 1

## Save to netCDF
If CLASS would be run without CTEM further variables should be set. See: https://cccma.gitlab.io/classic/basicInputs.html -> Required vegetation data


In [40]:
ds.to_netcdf(path_out+'/'+site+'_init_spinup_'+exp+'.nc')
ds.to_netcdf(path_out+'/rsfile_spinup_'+exp+'.nc')

In [41]:
!mkdir /home/lalandmi/eccc/classic-develop/outputFiles/SnowMIP/{site}/spinup_{exp}